File: src/surface-generation/biomes/chunks.ts
import { BiomeResolver } from "./biome-source"
import {
    Blender,
    ChunkPos,
    NoiseChunk,
    NoiseGeneratorSettings,
    NoiseSampler,
    QuartPos,
} from "./chunk-generator"
import { BlockPos, SectionPos } from "./pos"
import { Climate } from "./climate"
import { ChunkStatus } from "./chunk-status"
import { Biomes } from "./biomes"
import { Mth } from "./mth"
import { Blocks } from "./blocks"
import { Supplier } from "./consumer"
import { NoiseFiller } from "./noise/blended-noise"
import { FluidPicker } from "./aquifer"
import { Heightmap } from "./heightmap"

// chunk access

export class LevelHeightAccessor {
    getHeight(): number {
        return 384
    }

    getMinBuildHeight(): number {
        return -64
    }

    getMaxBuildHeight(): number {
        return this.getMinBuildHeight() + this.getHeight()
    }

    getSectionsCount(): number {
        return this.getMaxSection() - this.getMinSection()
    }

    getMinSection(): number {
        return SectionPos.blockToSectionCoord(this.getMinBuildHeight())
    }

    getMaxSection(): number {
        return SectionPos.blockToSectionCoord(this.getMaxBuildHeight() - 1) + 1
    }

    isOutsideBuildHeight(y: number): boolean
    isOutsideBuildHeight(pos: BlockPos): boolean
    isOutsideBuildHeight(y: number | BlockPos): boolean {
        if (typeof y === "number") {
            return y < this.getMinBuildHeight() || y >= this.getMaxBuildHeight()
        } else {
            const pos = y
            return this.isOutsideBuildHeight(pos.y)
        }
    }

    getSectionIndex(coord: number): number {
        return this.getSectionIndexFromSectionY(SectionPos.blockToSectionCoord(coord))
    }

    getSectionIndexFromSectionY(y: number): number {
        return y - this.getMinSection()
    }

    getSectionYFromSectionIndex(sectionIndex: number): number {
        return sectionIndex + this.getMinSection()
    }
}
export class LevelChunkSection {
    public static readonly SECTION_WIDTH = 16
    public static readonly SECTION_HEIGHT = 16
    public static readonly SECTION_SIZE = 4096
    public static readonly BIOME_CONTAINER_BITS = 2
    public static readonly STATES_CONTAINER_BITS = 4

    public static readonly BIOME_CONTAINER_SIZE = 1 << LevelChunkSection.BIOME_CONTAINER_BITS
    public static readonly STATES_CONTAINER_SIZE = 1 << LevelChunkSection.STATES_CONTAINER_BITS

    public readonly bottomBlockY: number
    private readonly biomes: Biomes[]
    private readonly states: Blocks[]

    private nonEmptyBlockCount: number

    constructor(y: number) {
        this.bottomBlockY = LevelChunkSection.getBottomBlockY(y)

        const biomesSize = LevelChunkSection.BIOME_CONTAINER_SIZE
        this.biomes = new Array(biomesSize * biomesSize * biomesSize) as Biomes[]
        this.biomes.fill(Biomes.PLAINS)

        const statesSize = LevelChunkSection.STATES_CONTAINER_SIZE
        this.states = new Array(statesSize * statesSize * statesSize) as Blocks[]
        this.states.fill(Blocks.AIR)

        this.nonEmptyBlockCount = 0
    }

    static getBottomBlockY(y: number): number {
        return y << 4
    }

    private static getBlockStateIndex(x: number, y: number, z: number): number {
        return (
            z * LevelChunkSection.STATES_CONTAINER_SIZE * LevelChunkSection.STATES_CONTAINER_SIZE +
            y * LevelChunkSection.STATES_CONTAINER_SIZE +
            x
        )
    }

    private static getBiomesIndex(x: number, y: number, z: number): number {
        return (
            z * LevelChunkSection.BIOME_CONTAINER_SIZE * LevelChunkSection.BIOME_CONTAINER_SIZE +
            y * LevelChunkSection.BIOME_CONTAINER_SIZE +
            x
        )
    }

    getBlockState(x: number, y: number, z: number): Blocks {
        const index = LevelChunkSection.getBlockStateIndex(x, y, z)
        return this.states[index]
    }

    setBlockState(x: number, y: number, z: number, block: Blocks, checked = true): Blocks {
        const index = LevelChunkSection.getBlockStateIndex(x, y, z)
        const prevBlock = this.states[index]
        this.states[index] = block

        if (prevBlock !== Blocks.AIR) {
            --this.nonEmptyBlockCount
        }

        if (block !== Blocks.AIR) {
            ++this.nonEmptyBlockCount
        }

        return prevBlock
    }

    getNoiseBiome(x: number, y: number, z: number): Biomes {
        const index = LevelChunkSection.getBiomesIndex(x, y, z)
        return this.biomes[index]
    }

    hasOnlyAir(): boolean {
        return this.nonEmptyBlockCount == 0
    }

    fillBiomesFromNoise(
        resolver: BiomeResolver,
        sampler: Climate.Sampler,
        offsetX: number,
        offsetZ: number
    ): void {
        const offsetY = QuartPos.fromBlock(this.bottomBlockY)

        for (let x = 0; x < 4; ++x) {
            for (let y = 0; y < 4; ++y) {
                for (let z = 0; z < 4; ++z) {
                    const index = LevelChunkSection.getBiomesIndex(x, y, z)

                    this.biomes[index] = resolver.getNoiseBiome(
                        offsetX + x,
                        offsetY + y,
                        offsetZ + z,
                        sampler
                    )
                }
            }
        }
    }
}

export abstract class ChunkAccess extends LevelHeightAccessor {
    private readonly sections: LevelChunkSection[]
    private readonly heightmaps: Map<Heightmap.Types, Heightmap>
    private noiseChunk: NoiseChunk

    constructor(
        readonly chunkPos: ChunkPos,
        private readonly levelHeightAccessor: LevelHeightAccessor
    ) {
        super()
        this.sections = new Array(levelHeightAccessor.getSectionsCount()) as LevelChunkSection[]
        this.heightmaps = new Map()

        ChunkAccess.replaceMissingSections(levelHeightAccessor, this.sections)
    }

    public getOrCreateHeightmapUnprimed(type: Heightmap.Types): Heightmap {
        return Mth.computeIfAbsent(this.heightmaps, type, type => new Heightmap(this, type))
    }

    private static replaceMissingSections(
        heightAccessor: LevelHeightAccessor,
        section: LevelChunkSection[]
    ): void {
        for (let i = 0; i < section.length; ++i) {
            if (section[i] == null) {
                section[i] = new LevelChunkSection(heightAccessor.getSectionYFromSectionIndex(i))
            }
        }
    }

    getNoiseBiome(x: number, y: number, z: number): Biomes {
        const minY = QuartPos.fromBlock(this.getMinBuildHeight())
        const maxY = minY + QuartPos.fromBlock(this.getHeight()) - 1
        const clampedY = Mth.clamp(y, minY, maxY)
        const sectionIndex = this.getSectionIndex(QuartPos.toBlock(clampedY))
        return this.sections[sectionIndex].getNoiseBiome(x & 3, clampedY & 3, z & 3)
    }

    abstract getStatus(): ChunkStatus

    fillBiomesFromNoise(resolver: BiomeResolver, sampler: Climate.Sampler): void {
        const chunkpos = this.getPos()
        const x = QuartPos.fromBlock(chunkpos.getMinBlockX())
        const z = QuartPos.fromBlock(chunkpos.getMinBlockZ())
        const heightAccessor = this.getHeightAccessorForGeneration()

        for (let y = heightAccessor.getMinSection(); y < heightAccessor.getMaxSection(); ++y) {
            const section = this.getSection(this.getSectionIndexFromSectionY(y))
            section.fillBiomesFromNoise(resolver, sampler, x, z)
        }
    }

    getSections() {
        return this.sections
    }

    getSection(sectionIndex: number): LevelChunkSection {
        return this.getSections()[sectionIndex]
    }

    getPos(): ChunkPos {
        return this.chunkPos
    }

    getMinBuildHeight(): number {
        return this.levelHeightAccessor.getMinBuildHeight()
    }

    getMaxBuildHeight(): number {
        return this.levelHeightAccessor.getMaxBuildHeight()
    }

    getHeight(): number {
        return this.levelHeightAccessor.getHeight()
    }

    getHighestSection(): LevelChunkSection | null {
        const sections = this.getSections()

        for (let i = sections.length - 1; i >= 0; --i) {
            const section = sections[i]
            if (!section.hasOnlyAir()) {
                return section
            }
        }

        return null
    }

    getHighestSectionPosition(): number {
        const levelchunksection = this.getHighestSection()
        return levelchunksection == null ? this.getMinBuildHeight() : levelchunksection.bottomBlockY
    }

    public getHeightAccessorForGeneration(): LevelHeightAccessor {
        return this
    }

    getOrCreateNoiseChunk(
        sampler: NoiseSampler,
        filler: Supplier<NoiseFiller>,
        settings: NoiseGeneratorSettings,
        fluidPicker: FluidPicker,
        blender: Blender
    ): NoiseChunk {
        if (this.noiseChunk == null) {
            this.noiseChunk = NoiseChunk.forChunk(
                this,
                sampler,
                filler,
                settings,
                fluidPicker,
                blender
            )
        }

        return this.noiseChunk
    }

    abstract getBlockState(pos: BlockPos): Blocks
}

export class ProtoChunk extends ChunkAccess {
    private status = ChunkStatus.EMPTY

    getStatus(): ChunkStatus {
        return this.status
    }

    getBlockState(pos: BlockPos): Blocks {
        const y = pos.y
        if (this.isOutsideBuildHeight(y)) {
            return Blocks.VOID_AIR
        } else {
            const section = this.getSection(this.getSectionIndex(y))
            return section.hasOnlyAir()
                ? Blocks.AIR
                : section.getBlockState(pos.x & 15, y & 15, pos.z & 15)
        }
    }
}

File: src/cpp/surface-generation/biomes/chunks.cpp
#include "chunks.hpp"
#include "chunk-generator.hpp"
#include "chunk-status.hpp"
#include "heightmap.hpp"
#include "mth.hpp"
#include "noise-chunk.hpp"

// LevelHeightAccessor

int32_t LevelHeightAccessor::getMaxBuildHeight() const {
    return this->getMinBuildHeight() + this->getHeight();
}

int32_t LevelHeightAccessor::getSectionsCount() const {
    return this->getMaxSection() - this->getMinSection();
}

int32_t LevelHeightAccessor::getMinSection() const {
    return SectionPos::blockToSectionCoord(this->getMinBuildHeight());
}

int32_t LevelHeightAccessor::getMaxSection() const {
    return SectionPos::blockToSectionCoord(this->getMaxBuildHeight() - 1) + 1;
}

bool LevelHeightAccessor::isOutsideBuildHeight(BlockPos const &pos) const {
    return this->isOutsideBuildHeight(pos.getY());
}

bool LevelHeightAccessor::isOutsideBuildHeight(int32_t y) const {
    return y < this->getMinBuildHeight() || y >= this->getMaxBuildHeight();
}

int32_t LevelHeightAccessor::getSectionIndex(int32_t y) const {
    return this->getSectionIndexFromSectionY(SectionPos::blockToSectionCoord(y));
}

int32_t LevelHeightAccessor::getSectionIndexFromSectionY(int32_t sectionY) const {
    return sectionY - this->getMinSection();
}

int32_t LevelHeightAccessor::getSectionYFromSectionIndex(int32_t sectionIndex) const {
    return sectionIndex + this->getMinSection();
}

// SimpleLevelHeightAccessor

int32_t SimpleLevelHeightAccessor::getHeight() const {
    return 384;
}

int32_t SimpleLevelHeightAccessor::getMinBuildHeight() const {
    return -64;
}

// LevelChunkSection

LevelChunkSection::LevelChunkSection(int32_t y) {
    this->_bottomBlockY = getBottomBlockY(y);

    this->states = vector<BlockState>(STATES_CONTAINER_SIZE * STATES_CONTAINER_SIZE * STATES_CONTAINER_SIZE);
    fill(this->states.begin(), this->states.end(), Blocks::AIR);

    this->biomes = vector<Biomes>(BIOME_CONTAINER_SIZE * BIOME_CONTAINER_SIZE * BIOME_CONTAINER_SIZE);
    fill(this->biomes.begin(), this->biomes.end(), Biomes::PLAINS);
}

BlockState LevelChunkSection::getBlockState(int32_t x, int32_t y, int32_t z) const {
    return this->states.at(getBlockStateIndex(x, y, z));
}

void LevelChunkSection::acquire() {
    // TODO ?
}

void LevelChunkSection::release() {
    // TODO ?
}

BlockState LevelChunkSection::setBlockState(int32_t x, int32_t y, int32_t z, BlockState blockState) {
    return this->setBlockState(x, y, z, blockState, true);
}

BlockState LevelChunkSection::setBlockState(int32_t x, int32_t y, int32_t z, BlockState blockState, bool checked) {
    BlockState prevBlockState;
    if (checked) {
        prevBlockState = this->states.at(getBlockStateIndex(x, y, z));
        this->states.at(getBlockStateIndex(x, y, z)) = blockState;
    } else {
        prevBlockState = this->states.at(getBlockStateIndex(x, y, z));
        this->states.at(getBlockStateIndex(x, y, z)) = blockState;
    }

    if (prevBlockState != Blocks::AIR) {
        --this->nonEmptyBlockCount;
    }

    if (blockState != Blocks::AIR) {
        ++this->nonEmptyBlockCount;
    }

    return prevBlockState;
}

bool LevelChunkSection::hasOnlyAir() const {
    return this->nonEmptyBlockCount == 0;
}

bool LevelChunkSection::isRandomlyTicking() const {
    return this->isRandomlyTickingBlocks() || this->isRandomlyTickingFluids();
}

bool LevelChunkSection::isRandomlyTickingBlocks() const {
    return this->tickingBlockCount > 0;
}

bool LevelChunkSection::isRandomlyTickingFluids() const {
    return this->tickingFluidCount > 0;
}

int32_t LevelChunkSection::bottomBlockY() const {
    return this->_bottomBlockY;
}

vector<BlockState> const &LevelChunkSection::getStates() const {
    return this->states;
}

vector<Biomes> const &LevelChunkSection::getBiomes() const {
    return this->biomes;
}

Biomes LevelChunkSection::getNoiseBiome(int32_t x, int32_t y, int32_t z) const {
    return this->biomes.at(getBiomesIndex(x, y, z));
}

void LevelChunkSection::fillBiomesFromNoise(shared_ptr<BiomeResolver> resolver, shared_ptr<Climate::Sampler> sampler,
                                            int32_t offsetX, int32_t offsetZ) {
    vector<Biomes> &biomes = this->biomes;
    // biomes->acquire();

    // try {
    int32_t offsetY = QuartPos::fromBlock(this->bottomBlockY());

    for (int32_t x = 0; x < BIOME_CONTAINER_SIZE; ++x) {
        for (int32_t y = 0; y < BIOME_CONTAINER_SIZE; ++y) {
            for (int32_t z = 0; z < BIOME_CONTAINER_SIZE; ++z) {
                biomes.at(getBiomesIndex(x, y, z)) =
                    resolver->getNoiseBiome(offsetX + x, offsetY + y, offsetZ + z, sampler);
            }
        }
    }
    /*
    } finally {
        biomes->release();
    }
    */
}

// ChunkAccess

ChunkAccess::ChunkAccess(ChunkPos const &chunkPos, LevelHeightAccessor const &levelHeightAccessor)
    : chunkPos(chunkPos), levelHeightAccessor(levelHeightAccessor) {
    this->noiseChunk = nullptr;
    this->isLightCorrect = false;
    this->sections = vector<LevelChunkSection>(levelHeightAccessor.getSectionsCount());
    this->heightmaps = std::map<HeightmapTypes, shared_ptr<Heightmap>>();

    replaceMissingSections(levelHeightAccessor, this->sections);

    objectCreated("ChunkAccess");
}

ChunkAccess::~ChunkAccess() {
    objectFreed("ChunkAccess");
}

void ChunkAccess::replaceMissingSections(LevelHeightAccessor const &heightAccessor,
                                         vector<LevelChunkSection> &sections) {
    for (int32_t sectionIndex = 0; sectionIndex < sections.size(); ++sectionIndex) {
        sections.at(sectionIndex) = LevelChunkSection(heightAccessor.getSectionYFromSectionIndex(sectionIndex));
    }
}

LevelChunkSection *ChunkAccess::getHighestSection() {
    vector<LevelChunkSection> &sections = this->getSections();

    for (int32_t sectionIndex = (int32_t)sections.size() - 1; sectionIndex >= 0; --sectionIndex) {
        LevelChunkSection &section = this->getSection(sectionIndex);
        if (!section.hasOnlyAir()) {
            return &section;
        }
    }

    return nullptr;
}

int32_t ChunkAccess::getHighestSectionPosition() {
    LevelChunkSection *section = this->getHighestSection();
    return section == nullptr ? this->getMinBuildHeight() : section->bottomBlockY();
}

vector<LevelChunkSection> &ChunkAccess::getSections() {
    return this->sections;
}

LevelChunkSection &ChunkAccess::getSection(int32_t sectionIndex) {
    return this->sections.at(sectionIndex);
}

shared_ptr<Heightmap> ChunkAccess::getOrCreateHeightmapUnprimed(HeightmapTypes type) {
    auto iterator = this->heightmaps.find(type);
    if (iterator == this->heightmaps.end()) {
        shared_ptr<Heightmap> heightmap = make_shared<Heightmap>(this->shared_from_this(), type);
        this->heightmaps.insert({type, heightmap});
        return this->heightmaps.at(type);
    } else {
        return iterator->second;
    }
}

bool ChunkAccess::hasPrimedHeightmap(HeightmapTypes type) const {
    return this->heightmaps.find(type) != this->heightmaps.end();
}

int32_t ChunkAccess::getHeight(HeightmapTypes type, int32_t x, int32_t z) {
    auto heightmapIterator = this->heightmaps.find(type);
    if (heightmapIterator == this->heightmaps.end()) {
        Heightmap::primeHeightmaps(this->shared_from_this(), {type});
    }
    shared_ptr<Heightmap> heightmap = this->heightmaps.at(type);
    return heightmap->getFirstAvailable(x & 15, z & 15) - 1;
}

ChunkPos const &ChunkAccess::getPos() const {
    return this->chunkPos;
}

int32_t ChunkAccess::getMinBuildHeight() const {
    return this->levelHeightAccessor.getMinBuildHeight();
}

int32_t ChunkAccess::getHeight() const {
    return this->levelHeightAccessor.getHeight();
}

shared_ptr<NoiseChunk> ChunkAccess::getOrCreateNoiseChunk(shared_ptr<NoiseSampler> sampler,
                                                          function<NoiseFiller(void)> filler,
                                                          NoiseGeneratorSettings const &settings,
                                                          shared_ptr<Aquifer::FluidPicker> fluidPicker,
                                                          Blender const &blender) {
    if (this->noiseChunk == nullptr) {
        this->noiseChunk =
            NoiseChunk::forChunk(this->shared_from_this(), sampler, filler, settings, fluidPicker, blender);
    }

    return this->noiseChunk;
}

Biomes ChunkAccess::getNoiseBiome(int32_t x, int32_t y, int32_t z) {
    int32_t minY = QuartPos::fromBlock(this->getMinBuildHeight());
    int32_t maxY = minY + QuartPos::fromBlock(this->getHeight()) - 1;
    int32_t clampedY = Mth::clamp(y, minY, maxY);
    int32_t sectionIndex = this->getSectionIndex(QuartPos::toBlock(clampedY));
    return this->getSection(sectionIndex).getNoiseBiome(x & 3, clampedY & 3, z & 3);
}

void ChunkAccess::fillBiomesFromNoise(shared_ptr<BiomeResolver> resolver, shared_ptr<Climate::Sampler> sampler) {
    ChunkPos const &chunkpos = this->getPos();
    int32_t x = QuartPos::fromBlock(chunkpos.getMinBlockX());
    int32_t z = QuartPos::fromBlock(chunkpos.getMinBlockZ());
    LevelHeightAccessor const &heightAccessor = this->getHeightAccessorForGeneration();

    for (int32_t y = heightAccessor.getMinSection(); y < heightAccessor.getMaxSection(); ++y) {
        LevelChunkSection &section = this->getSection(this->getSectionIndexFromSectionY(y));
        section.fillBiomesFromNoise(resolver, sampler, x, z);
    }
}

LevelHeightAccessor const &ChunkAccess::getHeightAccessorForGeneration() {
    return *this;
}

// ProtoChunk

ProtoChunk::ProtoChunk(ChunkPos const &chunkPos, LevelHeightAccessor const &levelHeightAccessor)
    : ChunkAccess(chunkPos, levelHeightAccessor), status(nullptr) {
}

BlockState ProtoChunk::getBlockState(BlockPos const &pos) const {
    int32_t y = pos.getY();
    if (this->isOutsideBuildHeight(y)) {
        return Blocks::VOID_AIR;
    } else {
        LevelChunkSection const &section = this->sections.at(this->getSectionIndex(y));
        return section.hasOnlyAir() ? Blocks::AIR : section.getBlockState(pos.getX() & 15, y & 15, pos.getZ() & 15);
    }
}

BlockState ProtoChunk::setBlockState(BlockPos const &pos, BlockState blockState, bool checked) {
    int32_t x = pos.getX();
    int32_t y = pos.getY();
    int32_t z = pos.getZ();
    if (y >= this->getMinBuildHeight() && y < this->getMaxBuildHeight()) {
        int32_t sectionIndex = this->getSectionIndex(y);
        if (this->getSection(sectionIndex).hasOnlyAir() && blockState == Blocks::AIR) {
            return blockState;
        } else {
            LevelChunkSection &section = this->getSection(sectionIndex);

            BlockState prevBlockState = section.setBlockState(x & 15, y & 15, z & 15, blockState);

            vector<HeightmapTypes> &heightmapsAfter = this->getStatus()->heightmapsAfter;
            vector<HeightmapTypes> createdHeightmaps = vector<HeightmapTypes>();

            for (HeightmapTypes &type : heightmapsAfter) {
                auto heightmap = this->heightmaps.find(type);
                if (heightmap == this->heightmaps.end()) {
                    createdHeightmaps.push_back(type);
                }
            }

            if (createdHeightmaps.size() > 0) {
                Heightmap::primeHeightmaps(this->shared_from_this(), createdHeightmaps);
            }

            for (HeightmapTypes &type : heightmapsAfter) {
                this->heightmaps.at(type)->update(x & 15, y, z & 15, blockState);
            }

            return prevBlockState;
        }
    } else {
        return Blocks::VOID_AIR;
    }
}

void ProtoChunk::setStatus(ChunkStatus* status) {
    this->status = status;
}

ChunkStatus* ProtoChunk::getStatus() {
    return this->status;
}
